### PR TITLE
CA-223406 - sr-scan with NFS4 broken

### PR DESCRIPTION
**Pull request by letsboogey**

### CODE REVIEW

**Comment by GaborApatiNagy** [comment]
I'll look into this tomorrow morning as agreed with @letsboogey

**Comment by GaborApatiNagy** [comment]
I think some more error handling is needed on the Location page. 

If I enter a Server name that is invalid/not accessible, there is no error message shown at all. Before this change, there used to be a popup window saying:
image

This version does not seem to display any error message: (It shows "Scanning..." for a few seconds only)
image

File: XenAdmin/Wizards/NewSRWizard_Pages/SrWizardHelpers.cs [hunk 1]
```diff
@@ -40,9 +40,15 @@ namespace XenAdmin.Wizards.NewSRWizard_Pages
 {
     public class SrWizardHelpers
     {
+        public static bool ValidateServerAddress(String address)
+        {
+            Regex re = new Regex(@"^[^/:*?<>|\\]+$");
```
Comment (GaborApatiNagy) [inline XenAdmin/Wizards/NewSRWizard_Pages/SrWizardHelpers.cs]: I think we have a better way to validate it, see: 
https://github.com/xenserver/xenadmin/blob/master/XenAdmin/Dialogs/OptionsPages/ConnectionOptionsPage.cs#L191

Uri.CheckHostName(string)

**Comment by GaborApatiNagy** [comment]
Thanks for the fixes @letsboogey 

I like that we keep the old screen for the previous versions and have the new layout for the new ones only.

I'm still at the initial UI testing - found these two issues, both with the new layout one:
* If the scan hasn't been successful, the subsequent controls shouldn't get enabled. (I see the error message is shown now, that's good.)
* I noticed that the label with the example gets updated if the text changes in the Share Name field. It should always match the version selection. I suggest you subscribe to the appropriate event of the the radio controls.

I'm still planning to look into the code in the meantime.

File: XenAdmin/Wizards/NewSRWizard.cs [hunk 2]
```diff
@@ -374,7 +384,7 @@ protected override void UpdateWizardContent(XenTabPage senderPage)
                 m_srWizardType.DeviceConfig = xenTabPageNfsIso.DeviceConfig;
                 SetCustomDescription(m_srWizardType, xenTabPageNfsIso.SrDescription);
             }
-            else if (senderPagetype == typeof(VHDoNFS))
+            else if (senderPagetype == typeof(VHDoNFS) || senderPagetype == typeof(VHDoNFSv2))
```
Comment (GaborApatiNagy) [inline XenAdmin/Wizards/NewSRWizard.cs]: Calling SetCustomDescription() for the new NFSv2 page is missing at 3 lines below

File: XenModel/Messages.resx [hunk 3]
```diff
@@ -5126,6 +5126,12 @@ Would you like to eject these ISOs before continuing?</value>
   <data name="EXAMPLE" xml:space="preserve">
     <value>Example: {0}</value>
   </data>
+  <data name="EXAMPLE_PATH_NFS3" xml:space="preserve">
+    <value>Example: /export/nfs</value>
+  </data>
+  <data name="EXAMPLE_PATH_NFS4" xml:space="preserve">
+    <value>Example: /nfs</value>
```
Comment (GaborApatiNagy) [inline XenModel/Messages.resx]: To keep this similar to the previous versions, I suggest we have "/path" here.

File: XenAdmin/Wizards/NewSRWizard_Pages/SrWizardHelpers.cs [hunk 4]
```diff
@@ -40,6 +40,21 @@ namespace XenAdmin.Wizards.NewSRWizard_Pages
 {
     public class SrWizardHelpers
     {
+        public static bool ValidateServerAddress(String address)
+        {
+            var serverNameType = Uri.CheckHostName(address);
+            if (serverNameType == UriHostNameType.Unknown || serverNameType == UriHostNameType.IPv6)
```
Comment (GaborApatiNagy) [inline XenAdmin/Wizards/NewSRWizard_Pages/SrWizardHelpers.cs]: When the then- and the else-statement is just a return true/false, it's usually better to just return the condition itself. I'm fine with this now, but it'd be nicer to have a one-liner here.

File: XenAdmin/Wizards/NewSRWizard_Pages/Frontends/VHDoNFSNew.cs [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+                {
+                    DisableControls();
+                    return;
+                }
+                                
+                NfsSharePathTextBox.Enabled = serverOptionsTextBox.Enabled = serverScanButton.Enabled = sharePathExampleLabel.Visible = true;
+                GetSupportedNfsVersionsAndSetUI(SrProbeAction.XMLFromErrorDescription);
+
+                //clear xml string otherwise it will persist across different probes and be reused if new one is not created, bypassing above if statement
+                SrProbeAction.XMLFromErrorDescription = null;
```
Comment (GaborApatiNagy) [inline XenAdmin/Wizards/NewSRWizard_Pages/Frontends/VHDoNFSNew.cs]: I can't see why we have a try-finally block here, but I'm fine with it.
However, if we have it around, this line is a typical example that should be in the finally block - to make sure it is always executed.
Therefore, I suggest we move this into the finally.

**Comment by letsboogey** [comment]
Thanks for the review and comments @GaborApatiNagy . I have made the requested changes

File: XenAdmin/Wizards/NewSRWizard.cs [hunk 6]
```diff
@@ -278,7 +280,12 @@ protected override void UpdateWizardContent(XenTabPage senderPage)
                 m_srWizardType = xenTabPageChooseSrType.SrWizardType;
 
                 if (m_srWizardType is SrWizardType_VhdoNfs)
-                    AddPage(xenTabPageVhdoNFS);
+                {
+                    if (Helpers.FalconOrGreater(xenConnection))
```
Comment (GaborApatiNagy) [inline XenAdmin/Wizards/NewSRWizard.cs]: I think this check should be IvernessOrGreater

**Comment by GaborApatiNagy** [comment]
Thanks @letsboogey , looks good now, merging

**Pull request merged**